### ISSUE TITLE
Closure

### ISSUE THREAD

**Issue by kayuksel** [vinbhaskara/AdamS#1]
Hello,

Thank you for publishing this interesting work. I can't wait to evaluate it on my use-cases. I have investigated main.py in this repository; yet, the closure doesn't calculate the loss again in there; as opposed to the following example from PyTorch website.

```
for input, target in dataset:
    def closure():
        optimizer.zero_grad()
        output = model(input)
        loss = loss_fn(output, target)
        loss.backward()
        return loss
    optimizer.step(closure)
```

My question is, if closure won't recalculate the loss, why it is necessary for AdamS? Which usage is proper for AdamS, the one from PyTorch website or the one in main.py? which is simply as follows:

```
        def closure(): return loss
        optimizer.zero_grad()
        loss.backward()
        optimizer.step(closure)
```

**Comment by kayuksel** [comment]
It also seems that I need to re-hypertune the learning rate for using AdamS, as the optimizer converges much slower when I experimented with that as a drop-in replacement for Adam.

**Comment by vinbhaskara** [comment]
Thank you for the question! 

Unlike Conjugate Gradient/LGBFS, the AdamX (X=UCB/CB/S) optimizers do not re-evaluate the loss multiple times per update. 

Only the current loss scalar value is necessary to estimate the running means and the variances. Therefore, ``closure`` is just used to pass the loss value for the current state of the network. So, it is one forward pass for one update.

Therefore, you can use either of the versions above. 

We used the latter one as that'd keep ``loss`` and ``output`` tensors in the training function scope for further use.

**Comment by vinbhaskara** [comment]
You can also try keeping the same learning rate you used to have in the original Adam but tune on the variance hyperparameter ``eta`` for AdamS for best convergence by quickly running it for ~3 epochs and picking the best one.

**Comment by kayuksel** [comment]
Thank you for your response. I have tried the values in the paper for the
eta value, will later try hypertuning it.

**Comment by kayuksel** [comment]
FYI, I am basically trying to employ it in the following Deep RL challenge I have created.

https://github.com/kayuksel/forex-rl-challenge

The repository also includes a gradient-free Augmented Random Search implementation, 
that seems to be much more successful than Adam version in converging to better policies.
